### ISSUE TITLE
How to do lemmatization?

### ISSUE THREAD

**Issue by fmv1992** [bowphs/SIGTYP-2024-hierarchical-transformers#1]
Hi, I've seen your models at https://huggingface.co/bowphs .

However, I'm at a loss on how I do lemmatization. For example, consider this text:

> ποικιλόθρον’ ἀθανάτ’ Ἀφρόδιτα,
> παῖ Δίος δολόπλοκε, λίσσομαί σε,
> μή μ’ ἄσαισι μηδ’ ὀνίαισι δάμνα,
> πότνια, θῦμον,

I'm a student of Ancient Greek and also a programmer (for context).

PS: thanks for your work :bow:

**Comment by fmv1992** [comment]
To make it super clear:

```
# https://github.com/bowphs/SIGTYP-2024-hierarchical-transformers/blob/main/lemmatization.py
from transformers import AutoTokenizer, T5ForConditionalGeneration

model = "bowphs/GreTa"
tokenizer = AutoTokenizer.from_pretrained(model)
model = T5ForConditionalGeneration.from_pretrained(model)

# inputs = ["lemmatize: ἀγαθή"]

inputs = ["lemmatize: <t_tok_beg> κατέβην <t_tok_end>"]

print("Tokenizing inputs...")
# Tokenize
encodings = tokenizer(inputs, return_tensors="pt", padding=False)

# Remove token_type_ids if present (T5 doesn't use them)
if "token_type_ids" in encodings:
    del encodings["token_type_ids"]

print("Generating outputs...")
# Generate
outputs = model.generate(**encodings)

print("Decoding outputs...")
# Decode
lemmas = tokenizer.batch_decode(outputs, skip_special_tokens=True)

print("Results:")
print(lemmas)
```

Results in:

```
Tokenizing inputs...
Generating outputs...
Decoding outputs...
Results:
[':  :  : > κατέβην <t']
```

**Comment by bowphs** [comment]
Hi and thanks for your interest in our work!

There are two important things to consider:

(1) First, the script that we provide lets you _fine-tune_ a pre-trained model for lemmatization, so you can control the data you input and get a model that performs your specific task of interest. Simply loading GreTa won't work, as GreTa is only pre-trained and was never fine-tuned to do lemmatization.

(2) Second, if you load a model that was fine-tuned on lemmatization, you still need to adhere to the exact formatting that was used during fine-tuning. Otherwise you'll confuse the model. If you just want to run inference with a reasonably performing model, I hope the code below helps:

```
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
from typing import List
import re

MODEL_PATH = "bowphs/lemmatization-demo"
MODEL_MAX_LENGTH = 512

lemmatizer_tokenizer = AutoTokenizer.from_pretrained(
    MODEL_PATH,
    model_max_length=MODEL_MAX_LENGTH,
)
lemmatizer_model = AutoModelForSeq2SeqLM.from_pretrained(MODEL_PATH)

def preprocess_text(text: str) -> List[str]:
    text = text.strip()
    text = re.sub("(?<! )(?=[.,!?()·;:])|(?<=[.,!?()·;:])(?! )", r" ", text)
    return text.split()

def lemmatize(tokens: List[str]) -> List[str]:
    """Lemmatize a list of tokens."""
    def construct_task(word_idx: int) -> str:
        return f"lemmatize: {' '.join(tokens[:word_idx])} <extra_id_0> {tokens[word_idx]} <extra_id_1> {' '.join(list(tokens[word_idx]))} <extra_id_2> {' '.join(tokens[word_idx+1:])}"

    predictions = [
        lemmatizer_tokenizer.decode(
            lemmatizer_model.generate(
                lemmatizer_tokenizer(construct_task(word_idx), return_tensors="pt")["input_ids"],
                max_length=20,
                num_beams=5,
                num_return_sequences=1,
                early_stopping=True,
            )[0],
            skip_special_tokens=True,
        )
        for word_idx in range(len(tokens))
    ]
    
    return predictions

if __name__ == "__main__":
    text = "Ἡροδότου Ἁλικαρνησσέος ἱστορίης ἀπόδεξις ἥδε, ὡς μήτε τὰ γενόμενα ἐξ ἀνθρώπων τῷ χρόνῳ ἐξίτηλα γένηται, μήτε ἔργα μεγάλα τε καὶ θωμαστά, τὰ μὲν Ἕλλησι τὰ δὲ βαρβάροισι ἀποδεχθέντα, ἀκλεᾶ γένηται, τά τε ἄλλα καὶ δι᾽ ἣν αἰτίην ἐπολέμησαν ἀλλήλοισι."
    tokens = preprocess_text(text)
    lemmata = lemmatize(tokens)
    
    for token, lemma in zip(tokens, lemmata):
        print(f"{token} -> {lemma}")
```


Let me know if you need anything else!

**Comment by fmv1992** [comment]
Hi, thanks for your detailed answer.

My only remark is that it would be nice if you provided a fully trained model that would fit most cases. Something more ready to use. But I know that might not be aligned with your research objectives.

Thanks again!
